happi/client.py

```diff
@@ -290,25 +335,69 @@ def all_devices(self):
         """
         A list of all contained devices
         """
-        return self.search()
+        return [res.device for res in self.search()]
+
+    def __getitem__(self, key):
+        'Get a device ID'
+        return SearchResult(client=self, metadata=self.backend.get_by_id(key))
+
+    def __iter__(self):
+        for info in self.backend.find({}):
+            yield info['_id']
+
+    def __len__(self):
+        return len(self.all_devices)
 
-    def search(self, start=0., end=None, as_dict=False, **kwargs):
+    def _get_search_results(self, items, *, wrap_cls=None):
+        '''
+        Return search results to the user, optionally wrapping with a class
+        '''
+        wrap_cls = wrap_cls or self._results_wrap_class
+        return [wrap_cls(client=self, metadata=info,
```

Comment:
Isn't this defeating the `yield` of the results?

Reply:
It removes the benefit of having the generator in the backend, yes. Should we have access to the generators in the user-facing API?